File: platform-frontend/src/models/aiMaterial.ts
import type { TelegramUser } from '@/models/profile'

export type AIMaterialContentType = 'prompt' | 'link' | 'agent'
export type AIMaterialKind = 'prompt' | 'skill' | 'library' | 'tutorial' | 'agent'
export type AIMaterialSort = 'new' | 'popular'

export interface AIMaterial {
  id: number
  authorId: number
  author?: TelegramUser
  title: string
  summary: string
  contentType: AIMaterialContentType
  materialKind: AIMaterialKind
  promptBody: string
  externalUrl: string
  agentConfig: string
  likesCount: number
  bookmarksCount: number
  commentsCount: number
  isHidden: boolean
  createdAt: string
  updatedAt: string
  tags: string[]
  liked: boolean
  bookmarked: boolean
}

export interface AIMaterialSearchResponse {
  items: AIMaterial[]
  total: number
}

export interface ToggleLikeResponse {
  liked: boolean
  likesCount: number
}

export interface ToggleBookmarkResponse {
  bookmarked: boolean
  bookmarksCount: number
}

export interface CreateAIMaterialRequest {
  title: string
  summary: string
  contentType: AIMaterialContentType
  materialKind: AIMaterialKind
  promptBody?: string
  externalUrl?: string
  agentConfig?: string
  tags: string[]
}

export interface AIMaterialFilters {
  kind?: AIMaterialKind | ''
  tag?: string
  q?: string
  sort?: AIMaterialSort
  mine?: boolean
  bookmarked?: boolean
  limit?: number
  offset?: number
}

export const AI_MATERIAL_KIND_LABELS: Record<AIMaterialKind, string> = {
  prompt: 'Промт',
  skill: 'Скилл',
  library: 'Библиотека',
  tutorial: 'Туториал',
  agent: 'Агент',
}

export const AI_MATERIAL_KIND_OPTIONS: { value: AIMaterialKind, label: string }[] = [
  { value: 'prompt', label: AI_MATERIAL_KIND_LABELS.prompt },
  { value: 'skill', label: AI_MATERIAL_KIND_LABELS.skill },
  { value: 'library', label: AI_MATERIAL_KIND_LABELS.library },
  { value: 'tutorial', label: AI_MATERIAL_KIND_LABELS.tutorial },
  { value: 'agent', label: AI_MATERIAL_KIND_LABELS.agent },
]

export const AI_MATERIAL_CONTENT_TYPE_LABELS: Record<AIMaterialContentType, string> = {
  prompt: 'Текст промта',
  link: 'Ссылка',
  agent: 'Конфиг агента',
}

export const AI_MATERIAL_CONTENT_TYPE_OPTIONS: { value: AIMaterialContentType, label: string, description: string }[] = [
  { value: 'prompt', label: 'Текст промта', description: 'Промт или инструкция, которую можно скопировать одной кнопкой' },
  { value: 'link', label: 'Ссылка', description: 'Внешний ресурс — Notion, GitHub, Telegraph и т.д.' },
  { value: 'agent', label: 'Конфиг агента', description: 'JSON/YAML конфиг готового AI-агента' },
]

export const AI_MATERIAL_LIMITS = {
  titleMin: 3,
  titleMax: 120,
  summaryMin: 30,
  summaryMax: 500,
  promptMax: 50_000,
  agentMax: 50_000,
  urlMax: 2048,
  tagsMax: 5,
  tagLenMax: 40,
} as const

File: backend/internal/models/ai_material.go
package models

import "time"

type AIMaterialContentType string
type AIMaterialKind string

const (
	AIMaterialContentTypePrompt AIMaterialContentType = "prompt"
	AIMaterialContentTypeLink   AIMaterialContentType = "link"
	AIMaterialContentTypeAgent  AIMaterialContentType = "agent"

	AIMaterialKindPrompt   AIMaterialKind = "prompt"
	AIMaterialKindSkill    AIMaterialKind = "skill"
	AIMaterialKindLibrary  AIMaterialKind = "library"
	AIMaterialKindTutorial AIMaterialKind = "tutorial"
	AIMaterialKindAgent    AIMaterialKind = "agent"

	AIMaterialMaxTags        = 5
	AIMaterialMaxTagLen      = 40
	AIMaterialMinTitleLen    = 3
	AIMaterialMaxTitleLen    = 120
	AIMaterialMinSummaryLen  = 30
	AIMaterialMaxSummaryLen  = 500
	AIMaterialMaxURLLen      = 2048
	AIMaterialMaxPromptBody  = 50_000
	AIMaterialMaxAgentConfig = 50_000
)

type AIMaterial struct {
	Id             int64                 `json:"id" gorm:"primaryKey"`
	AuthorId       int64                 `json:"authorId" gorm:"column:author_id;not null"`
	Author         *Member               `json:"author,omitempty" gorm:"foreignKey:AuthorId"`
	Title          string                `json:"title" gorm:"size:120;not null"`
	Summary        string                `json:"summary" gorm:"not null"`
	ContentType    AIMaterialContentType `json:"contentType" gorm:"column:content_type;size:16;not null"`
	MaterialKind   AIMaterialKind        `json:"materialKind" gorm:"column:material_kind;size:16;not null"`
	PromptBody     string                `json:"promptBody" gorm:"column:prompt_body;default:''"`
	ExternalURL    string                `json:"externalUrl" gorm:"column:external_url;size:2048;default:''"`
	AgentConfig    string                `json:"agentConfig" gorm:"column:agent_config;default:''"`
	LikesCount     int                   `json:"likesCount" gorm:"column:likes_count;default:0"`
	BookmarksCount int                   `json:"bookmarksCount" gorm:"column:bookmarks_count;default:0"`
	CommentsCount  int                   `json:"commentsCount" gorm:"column:comments_count;default:0"`
	IsHidden       bool                  `json:"isHidden" gorm:"column:is_hidden;default:false"`
	CreatedAt      time.Time             `json:"createdAt" gorm:"column:created_at;autoCreateTime"`
	UpdatedAt      time.Time             `json:"updatedAt" gorm:"column:updated_at;autoUpdateTime"`

	// Заполняются сервисом, не маппятся на колонки.
	Tags       []string `json:"tags" gorm:"-:all"`
	Liked      bool     `json:"liked" gorm:"-:all"`
	Bookmarked bool     `json:"bookmarked" gorm:"-:all"`
}

func (AIMaterial) TableName() string {
	return "ai_materials"
}

type AIMaterialTag struct {
	MaterialId int64  `gorm:"primaryKey;column:material_id"`
	Tag        string `gorm:"primaryKey;size:40;column:tag"`
}

func (AIMaterialTag) TableName() string {
	return "ai_material_tags"
}

type AIMaterialLike struct {
	MaterialId int64     `gorm:"primaryKey;column:material_id"`
	MemberId   int64     `gorm:"primaryKey;column:member_id"`
	CreatedAt  time.Time `gorm:"column:created_at;autoCreateTime"`
}

func (AIMaterialLike) TableName() string {
	return "ai_material_likes"
}

type AIMaterialBookmark struct {
	MaterialId int64     `gorm:"primaryKey;column:material_id"`
	MemberId   int64     `gorm:"primaryKey;column:member_id"`
	CreatedAt  time.Time `gorm:"column:created_at;autoCreateTime"`
}

func (AIMaterialBookmark) TableName() string {
	return "ai_material_bookmarks"
}

type AIMaterialComment struct {
	Id         int64     `json:"id" gorm:"primaryKey"`
	MaterialId int64     `json:"materialId" gorm:"column:material_id;not null"`
	AuthorId   int64     `json:"authorId" gorm:"column:author_id;not null"`
	Author     *Member   `json:"author,omitempty" gorm:"foreignKey:AuthorId"`
	Body       string    `json:"body" gorm:"not null"`
	LikesCount int       `json:"likesCount" gorm:"column:likes_count;default:0"`
	IsHidden   bool      `json:"isHidden" gorm:"column:is_hidden;default:false"`
	CreatedAt  time.Time `json:"createdAt" gorm:"column:created_at;autoCreateTime"`
	UpdatedAt  time.Time `json:"updatedAt" gorm:"column:updated_at;autoUpdateTime"`

	// Заполняется сервисом — флаг «текущий viewer лайкнул».
	Liked bool `json:"liked" gorm:"-:all"`
}

func (AIMaterialComment) TableName() string {
	return "ai_material_comments"
}

type AIMaterialCommentLike struct {
	CommentId int64     `gorm:"primaryKey;column:comment_id"`
	MemberId  int64     `gorm:"primaryKey;column:member_id"`
	CreatedAt time.Time `gorm:"column:created_at;autoCreateTime"`
}

func (AIMaterialCommentLike) TableName() string {
	return "ai_material_comment_likes"
}

type CreateAIMaterialRequest struct {
	Title        string                `json:"title"`
	Summary      string                `json:"summary"`
	ContentType  AIMaterialContentType `json:"contentType"`
	MaterialKind AIMaterialKind        `json:"materialKind"`
	PromptBody   string                `json:"promptBody"`
	ExternalURL  string                `json:"externalUrl"`
	AgentConfig  string                `json:"agentConfig"`
	Tags         []string              `json:"tags"`
}

type UpdateAIMaterialRequest = CreateAIMaterialRequest

func IsValidAIMaterialContentType(v AIMaterialContentType) bool {
	switch v {
	case AIMaterialContentTypePrompt, AIMaterialContentTypeLink, AIMaterialContentTypeAgent:
		return true
	}
	return false
}

func IsValidAIMaterialKind(v AIMaterialKind) bool {
	switch v {
	case AIMaterialKindPrompt, AIMaterialKindSkill, AIMaterialKindLibrary, AIMaterialKindTutorial, AIMaterialKindAgent:
		return true
	}
	return false
}
